src/Beman/Optional26/tests/optional_constexpr.t.cpp

```diff
@@ -0,0 +1,1744 @@
+};
+} // namespace
+
+TEST(OptionalConstexprTest, Constructors) {
+    constexpr beman::optional26::optional<int> i1;
+    constexpr beman::optional26::optional<int> i2{beman::optional26::nullopt};
+
+    constexpr int                              i  = 0;
+    constexpr beman::optional26::optional<int> i3 = i;
+    (void)i3;
+
+    constexpr beman::optional26::optional<empty> e1;
+    constexpr beman::optional26::optional<int>   e2{beman::optional26::nullopt};
+
+    constexpr empty                              e{};
+    constexpr beman::optional26::optional<empty> e3 = e;
+    (void)e3;
+}
+
+TEST(OptionalConstexprTest, Constructors2) {
+    constexpr beman::optional26::optional<int> o1;
+    EXPECT_TRUE(!o1);
+
+    constexpr beman::optional26::optional<int> o2 = beman::optional26::nullopt;
+    EXPECT_TRUE(!o2);
+
+    constexpr beman::optional26::optional<int> o3 = 42;
+    EXPECT_TRUE(*o3 == 42);
+
+    constexpr beman::optional26::optional<int> o4 = o3;
+    EXPECT_TRUE(*o4 == 42);
+
+    constexpr beman::optional26::optional<int> o5 = o1;
+    EXPECT_TRUE(!o5);
+
+    constexpr beman::optional26::optional<int> o6 = std::move(o3);
+    EXPECT_TRUE(*o6 == 42);
+
+    constexpr beman::optional26::optional<short> o7 = 42;
+    EXPECT_TRUE(*o7 == 42);
+
+    constexpr beman::optional26::optional<int> o8 = o7;
+    EXPECT_TRUE(*o8 == 42);
+
+    constexpr beman::optional26::optional<int> o9 = std::move(o7);
+    EXPECT_TRUE(*o9 == 42);
+
+    {
+        constexpr beman::optional26::optional<int&> o;
+        EXPECT_TRUE(!o);
+
+        constexpr beman::optional26::optional<int&> oo = o;
+        EXPECT_TRUE(!oo);
+    }
+
+    {
+        constexpr static auto                              i = 42;
+        constexpr beman::optional26::optional<const int&> o = i;
+        EXPECT_TRUE(o);
+        EXPECT_TRUE(*o == 42);
+
+        constexpr beman::optional26::optional<const int&> oo = o;
+        EXPECT_TRUE(oo);
+        EXPECT_TRUE(*oo == 42);
+    }
+
+}
+
+TEST(OptionalConstexprTest, Constructors3) {
+    constexpr beman::optional26::optional<int> ie;
+    constexpr beman::optional26::optional<int> i4 = ie;
+    EXPECT_FALSE(i4);
+
+    constexpr base                              b{1};
+    constexpr derived                           d(1, 2);
+    constexpr beman::optional26::optional<base> b1{b};
+    constexpr beman::optional26::optional<base> b2{d};
+
+    constexpr beman::optional26::optional<derived> d2{d};
+    constexpr beman::optional26::optional<base>    b3 = d2;
+    constexpr beman::optional26::optional<base>    b4{d2};
+}
+
+namespace {
+class NoDefault {
+    int v_;
+
+  public:
+    constexpr NoDefault(int v) : v_(v) {}
+};
+} // namespace
+
+TEST(OptionalConstexprTest, NonDefaultConstruct) {
+    constexpr NoDefault                              i = 7;
+    constexpr beman::optional26::optional<NoDefault> v1{};
+    constexpr beman::optional26::optional<NoDefault> v2{i};
+}
+
+TEST(OptionalConstexprTest, Triviality) {
```

Comment:
Can we reduce the duplication between optional.t.cpp and optional_constexpr.t.cpp?


Check https://gist.githubusercontent.com/neatudarius/1ded802e47996445ed4a731d10f7c4fc/raw/36448222892a0fe403ea39c5f0eb2c0dc89e9097/gistfile1.txt, e.g. `TEST(OptionalConstexprTest, Triviality) ` vs `TEST(OptionalTest, Triviality) `

Reply:
Factoring out repeated nonce types is something we should definitely do. Testing constexpr  is tricky, though. You have to call it in a consteval context, and that doesn't necessarily tell you if the runtime evaluation is correct, because they can be effectively unrelated. 
X macros or preprocessor tricks? I don't have a good idea off-hand for the parallelism.

Reply:
Ok, we can defer this task.